src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,298 @@
+
+     This class is updatable at run-time, allowing other libraries to add their
+     options own configuration options and sub-parsers. Sub-parsers allow
+     options groups to exist, e.g. the group "snowflake.cli.output" could have
+     2 options in it: debug (boolean flag) and format (a string like "json", or
+     "csv").
+
+    When a ConfigParser tries to retrieve ConfigOptions' value the _root_parser
+    will read and cache the TOML file from the file it's pointing at, afterwards
+    updating the read cache can be forced by calling read_config.
+
+    Attributes:
+        name: The name of the ConfigParser. Used for nesting and emitting
+          useful error messages.
+        file_path: Path to the file where this and all child ConfigParsers
+          should read their values out of. Can be omitted for all child
+          parsers.
+        conf_file_cache: Cache to store what we read from the TOML file.
+        _sub_parsers: List of ConfigParsers that are nested under us.
+        _options: List of ConfigOptions that are our children.
+        _root_parser: Reference to root parser. Used to efficiently propagate to
+          child options.
+        _nest_path: The names of the ConfigParsers that this parser is nested
+          under. Used to efficiently propagate to child options.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        """Create a new ConfigParser.
+
+        Args:
+            name: Name of this ConfigParser.
+            file_path: File this parser should read values from. Can be omitted
+              for all child parsers.
+        """
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
+        self._options: dict[str, ConfigOption] = dict()
+        self._sub_parsers: dict[str, ConfigParser] = dict()
+        # Dictionary to cache read in config file
+        self.conf_file_cache: TOMLDocument | None = None
+        # Information necessary to be able to nest elements
+        #  and add options in O(1)
+        self._root_parser: ConfigParser = self
+        self._nest_path = [name]
+
+    def read_config(
+        self,
+    ) -> None:
+        """Read and cache config file."""
+        if self.file_path is None:
+            raise ConfigParserError(
+                "ConfigParser is trying to read config file, but it doesn't have one"
+            )
+        LOGGER.debug(f"reading configuration file from {str(self.file_path)}")
+        try:
+            self.conf_file_cache = tomlkit.parse(self.file_path.read_text())
+        except Exception as e:
+            raise ConfigSourceError(
+                f"An unknown error happened while loading '{str(self.file_path)}"
+                f"', please see the error: {e}"
+            )
+
+    def add_option(
+        self,
+        *,
+        option_cls: type[ConfigOption] = ConfigOption,
```

Comment:
My reasoning is that since the class is pluggable with `option_cls` I think using keyword only arguments will save developers from a ton of headache.
In other parts of this file I also found that I couldn't decide on an intuitive way to order arguments.
I definitely welcome your feedback and wouldn't be opposed to another discussion about this. The only thing I ask is that we have this meeting this week!

Reply:
> all arguments for public APIs must be keyword-only
I don't agree with this. I think it's appropriate this time because the target audience is Snowflake developers. I'd definitely not do something like this with Public APIs that are targeted at end-users.

Reply:
> I don't agree with this. I think it's appropriate this time because the target audience is Snowflake developers for this. I'd definitely not do something like this with Public APIs that are targeted at end-users.

Fair enough.  I was curious because I've seen this principle described in other proposed internal style guides.  I disagreed with it as a general principle there and just wanted to call it out here.  Since it's a specific choice rather than a general principle, it's your call!